src/gui/TotpSetupDialog.cpp

```diff
@@ -43,51 +45,92 @@ TotpSetupDialog::~TotpSetupDialog()
 
 void TotpSetupDialog::saveSettings()
 {
+    // Secret key sanity check
+    auto key = m_ui->seedEdit->text().toLatin1();
+    auto sanitizedKey = Base32::sanitizeInput(key);
+    if (sanitizedKey != key) {
+        MessageBox::information(this,
+                                "Invalid TOTP Secret",
+                                "You have entered an invalid secret key. The key must be in Base32 format.\n"
+                                "Example: JBSWY3DPEHPK3PXP");
+        return;
+    }
+
     QString encShortName;
     uint digits = Totp::DEFAULT_DIGITS;
     uint step = Totp::DEFAULT_STEP;
+    Totp::Algorithm algorithm = Totp::DEFAULT_ALGORITHM;
+    Totp::StorageFormat format = Totp::DEFAULT_FORMAT;
 
     if (m_ui->radioSteam->isChecked()) {
         digits = Totp::STEAM_DIGITS;
         encShortName = Totp::STEAM_SHORTNAME;
     } else if (m_ui->radioCustom->isChecked()) {
+        algorithm = static_cast<Totp::Algorithm>(m_ui->algorithmComboBox->currentData().toInt());
         step = m_ui->stepSpinBox->value();
-        if (m_ui->radio8Digits->isChecked()) {
-            digits = 8;
-        } else if (m_ui->radio7Digits->isChecked()) {
-            digits = 7;
+        digits = m_ui->digitsSpinBox->value();
+    }
+
+    auto settings = m_entry->totpSettings();
+    if (settings) {
+        if (key.isEmpty()) {
+            auto answer = MessageBox::question(this,
+                                               "Confirm Remove TOTP Settings",
+                                               "Are you sure you want to delete TOTP settings for this entry?",
+                                               MessageBox::Delete | MessageBox::Cancel);
+            if (answer == MessageBox::Cancel) {
+                return;
+            }
+        }
+
+        format = settings->format;
+        if (format == Totp::StorageFormat::LEGACY && m_ui->radioCustom->isChecked()) {
+            auto answer = MessageBox::question(
+                this,
+                "Upgrade TOTP Settings",
+                "You are currently using legacy TOTP settings which do not fully support custom settings.\n"
+                "Would you like to upgrade to the latest format?",
+                MessageBox::Yes | MessageBox::No,
+                MessageBox::No);
```

Comment:
Not translatable either. I am also quite afraid that this might confuse and unsettle users again (like our "legacy key file" dialogue did). What could go wrong if we just upgrade silently? Why not simply create a history snapshot before saving the upgraded settings? And then if we still don't trust our own programming skills, we can show a purely informative message saying "hey there, your TOTP settings were upgraded. Please check that everything is in order, you can revert the changes in the history panel".

Reply:
I removed the prompt, I had the same thoughts. Since we are thinking the same thing it is best to just implicitly "upgrade" and don't event bother the user. The only downfall is that legacy TOTP plugins *may* not recognize the OTP URL format. But if you are using custom TOTP settings it is likely they won't recognize those either.